scipy/stats/_binned_statistic.py

```diff
@@ -578,7 +578,11 @@ def binned_statistic_dd(sample, values, statistic='mean',
         dedges = [np.diff(edges[i]) for i in builtins.range(Ndim)]
         binnumbers = binned_statistic_result.binnumber
 
-    result = np.empty([Vdim, nbin.prod()], float)
+    if np.iscomplexobj(values):
+        result = np.empty([Vdim, nbin.prod()], values.dtype)
+    else:
+        # Use a float64 accumulator to avoid integer overflow
+        result = np.empty([Vdim, nbin.prod()], float)
```

Comment:
It's not so much the `dtype` of `values` that matters, but of `statistic(values)`. If the statistic were `np.count`, for instance, we wouldn't want `dtype` to be complex.

Reply:
I've taken this advice and made the necessary changes for this to happen. Unfortunately had to change _calc_binned_statistic to have to return a value rather than operate in-place to account for the dtype change when it first encounters the result of `statistic(values)` as being complex